tests/func/test_remote_dir.py

```diff
@@ -38,7 +51,54 @@ def remote():
         yield remote
 
 
-def test_isdir(remote):
+@pytest.fixture
+def remote_gs():
+    """Returns a RemoteGS connected to a bucket with the following structure:
+        bucket
+        ├── data
+        │  ├── alice
+        │  ├── alpha
+        │  └── subdir
+        │     ├── 1
+        │     ├── 2
+        │     └── 3
+        ├── empty_dir
+        ├── empty_file
+        └── foo
+    """
+    prefix = str(uuid.uuid4())
+    REMOTE_URL = "gs://dvc-test/" + prefix
+    remote = RemoteGS(None, {"url": REMOTE_URL})
+    teardowns = []
+
+    def put_object(file, content):
+        create_object_gs(remote.gs, "dvc-test", prefix + "/" + file, content)
+        teardowns.append(lambda: remote.remove(remote.path_info / file))
+
+    put_object("empty_dir/", "")
+    put_object("empty_file", "")
+    put_object("foo", "foo")
+    put_object("data/alice", "alice")
+    put_object("data/alpha", "alpha")
+    put_object("data/subdir/1", "1")
+    put_object("data/subdir/2", "2")
+    put_object("data/subdir/3", "3")
+
+    yield remote
+
+    for teardown in teardowns:
+        teardown()
+
+
+remote_parameterized = pytest.mark.parametrize(
+    "remote_name", [pytest.param("remote_gs", marks=test_gs), "remote_s3"]
+)
```

Comment:
We are starting to duplicate whatever we have in `tests/func/test_api.py`. Let's extract remote definitions from there to `tests/remotes.py` and supplement that with whatever you need here. E.g. with `.get_remote()` and/or `.put_object()`.

Then make a `remote` fixture:
```python
@pytest.fixture
def remote(request):
    if not request.param.should_test():
        raise pytest.skip()
    return request.param.get_remote()
```

Reply:
Ah, did not know that. Will do.

Reply:
@Suor, I have refactored these, but, I went with using `contextmanager` so not to create any track any *states*. Please review again.